### ISSUE TITLE
Best-practice stateful testing of data structures with non-trivial construction parameters

### ISSUE THREAD

**Issue by glowthrower** [HypothesisWorks/hypothesis#2860]
I've recently been experimenting with the [stateful testing mechanisms](https://hypothesis.readthedocs.io/en/latest/stateful.html). They're proving to be extremely useful and I've already been able to track down some really intricate bugs, but I'm interested in making sure that I'm using the tools to their best extent. As such, I have some minor questions about "correct" and elegant usage that I was hoping someone could provide some guidance on.

The basic setting I'm working in is that of  a stateful data structure, with a constructor that takes at least one non-trivial parameter. As an artificial but illustrative example, consider the following stack implementation:
```python
from typing import Any, Iterable, Optional

class BrokenStack:
    def __init__(self, initial_values: Optional[Iterable[Any]] = None) -> None:
        self._stack = list(initial_values) if initial_values else []
        self._initial_size = len(self._stack)
        self._push_count = 0
        self._pop_count = 0

    def push(self, value: Any) -> None:
        self._stack.append(value)
        self._push_count += 1

    def pop(self) -> Any:
        value = self._stack.pop()
        self._pop_count += 1
        if (self._initial_size > 10) and (
            (self._pop_count + self._push_count) == self._initial_size
        ):
            raise RuntimeError("Construction-parameter dependent bug.")
        return value

    def can_pop(self) -> bool:
        return len(self._stack) > 0
```
I can easily write a `RuleBasedStateMachine` that will explore the state space of this data structure:
```python
import hypothesis.strategies as hp_st
import hypothesis.stateful as hp_state


class StackTester(hp_state.RuleBasedStateMachine):
    def __init__(self) -> None:
        super().__init__()
        self.stack = BrokenStack()

    @hp_state.rule(value=hp_st.integers())
    def push_something(self, value: int) -> None:
        self.stack.push(value)

    @hp_state.precondition(lambda self: self.stack.can_pop())
    @hp_state.rule()
    def pop_something(self) -> None:
        self.stack.pop()


StackTest = StackTester.TestCase
```
Note that I have simply called the constructor for `BrokenStack` with no arguments. Running this test as given produces a pass, as the bug is never triggered. If I modify the constructor call to, e.g. `BrokenStack(range(11))`, then a failure case is found in short order. So far so good, and indeed a tidy illustration of the power of the stateful testing mechanism. But either way, I'm still testing only one possible choice of construction parameter out of a potential universe.

The obvious solution is to use standard hypothesis strategies to provide the input values for the constructor, or (in more complicated cases) write composite strategies to generate an initial instance of the data structure under testing. It seems that I can use `@initialize` rules to do exactly this, like so:
```python
class BetterStackTester(hp_state.RuleBasedStateMachine):
    def __init__(self) -> None:
        super().__init__()
        self.stack: Optional[BrokenStack] = None

    @hp_state.initialize(
        initial_values=hp_st.one_of(hp_st.none(), hp_st.lists(hp_st.integers()))
    )
    def construct_stack(self, initial_values: Optional[Iterable[Any]]) -> None:
        self.stack = BrokenStack(initial_values)

    @hp_state.rule(value=hp_st.integers())
    def push_something(self, value: int) -> None:
        assert isinstance(self.stack, BrokenStack)
        self.stack.push(value)

    @hp_state.precondition(lambda self: self.stack.can_pop())
    @hp_state.rule()
    def pop_something(self) -> None:
        assert isinstance(self.stack, BrokenStack)
        self.stack.pop()


StackTest = BetterStackTester.TestCase
```
This works -- the bug is discovered easily. And on the face of it, it's a tidy way to do it -- the only extra boilerplate required is the explicit `assert isinstance(self.stack, BrokenStack)` lines at the beginning of each rule, which keep my type-checker happy. But there are a couple of issues here that make me wonder if this couldn't be improved.

1) The language of the documentation is a little unclear as to what contracts are enforced here. The docs state that `@initialize` rules "are guaranteed to be run at most once at the beginning of a run (i.e. before any normal rule is called)", which could be read as leaving open the possibility that some or all such rules will *not* be called before entering the "main" run. That would mean that every normal rule needs to consider the possibility that it is being applied to an uninitialised state machine. This can be treated by placing a precondition decorator on top of every `@rule`, but in my real-world cases, I'm already making heavy use of preconditions, and adding something like `self.was_initialised and ...` to each of them feels error-prone and overly verbose. This may not actually be required: the [underlying source code](https://github.com/HypothesisWorks/hypothesis/blob/d76062140369488ccff387d6bc78cf2c5995d80b/hypothesis-python/src/hypothesis/stateful.py#L132) seems to only ever execute normal rules if all initialisation rules have been executed, but I'm not sure if I can rely on that going forwards. What's the opinion of the devs?

2) Any `@invariant` tests that I specify must also hold directly after state machine construction, before the execution of any `@initialize` rule, and so must consider the case of an uninitialised machine. Again, it seems that I can use `@precondition` decorators to enforce e.g. `self.was_initialised`, but this also feels a bit "wrong", since the notion of guarding an invariant seems contrary to the definition of invariance. Is there a way to restrict invariants to only being applied during the rule interleaving after initialisation, without having to DIY?

3) It's not clear to me if I'm trying to jam "too much" testing into each test case, especially if the strategy used to construct the data structure is more complex than the toy example I've given here. Even if I just fix constructor arguments, my stateful tests are already significantly more expensive than standard property based tests (seconds rather than milliseconds), and I'm not sure if the extra effort to cover the manifold possible construction possibilities is detracting from the amount of effort available to explore the state space. Could I structure things more effectively, or in a way that is more amenable to test parallelisation?

Any ideas or comments would be very welcome!  (And, by the way, many, many thanks in general to all involved in the hypothesis project!)

**Comment by Zalathar** [comment]
From looking at #1216 and #1291, it seems that the intent was to run *all* initializers before *any* regular rules. And one of the explicit use cases was calling a non-trivial constructor with generated data before poking the resulting object's stateful methods, which only makes sense if you have that ordering guarantee.

[The `@initialize` docs](https://hypothesis.readthedocs.io/en/latest/stateful.html#hypothesis.stateful.initialize) also have this to say:

> All initialize decorated methods will be called before any rule decorated methods, in an arbitrary order.

So that all seems pretty reassuring that you're on the right track.

**Comment by Zalathar** [comment]
For the `@initialize`/`@invariant` interaction, this seems to just be a combination of features that wasn't really considered when initializers were added, so the current behaviour is probably just an accidental consequence of initializers being treated as a special subset of rules under the hood.

Automatically skipping invariant checks until all initializers are complete seems much more sensible than the current behaviour, though I'm not sure whether or not this needs to be handled as a breaking change.

**Comment by glowthrower** [comment]
Thanks for the swift feedback :smiley: Glad to hear that I'm not going too far sideways.

I definitely agree that invariant checks being deferred until after initialisation is complete would be more natural for cases like the one I've outlined. I wasn't sure from the docs if the intended use case for `@initialize` was more related to populating bundles than to initialising the state machine itself; in that kind of a case, the use of invariants might conceivably make more sense.

(As a general side comment, I feel that the docs might slightly overemphasise the use of bundles -- it took me a couple of read-throughs to work out that they were more of a convenience data structure for separating state transitions and data generation, rather than an integral part of the API.)

**Comment by Zac-HD** [comment]
That's a beautifully clear writeup; thanks - and to agree with Zalathar, we certainly intend to support the stuff you're doing so the feedback on docs and design are very valuable 😁 

1. I think we should rephrase the docs to clarify this, for example by adding:

   > An `@initialize` method will only be skipped if one of the earlier methods raises an exception, in which case the machine calls `self.teardown()` and exits.

2. Agreed - it would make sense to skip `@invariant()` methods during initialization by default.
   This is certainly a change in behaviour and arguably documented (does "run after every rule" include after initialize rules?), but I'm comfortable disregarding [Hyrum's law](https://www.hyrumslaw.com/) here because invariants [should](https://xkcd.com/1172/) *really* be pure functions.  Calling them less often is therefore not a breaking change.
   I'd be open to adding a way to opt back in to the old behaviour though, e.g. `@invariant(even_incomplete=True)`, if only someone can think of a good argument name.

3. I don't have any particular advice here beyond the observation that varying setup parameters means your tests are capable of finding more bugs.  It's the usual tradeoff between example-based and property-based testing, so there's no universal answer but I usually favor the extra variations (which are pretty cheap if you're already going to run the state machine).

I tend to feel the same way about our docs emphasising bundles, but until someone volunteers to improve them that's what we've got.  https://hypothesis.works/articles/how-not-to-die-hard-with-hypothesis/ is a nice intro, at least, and https://hypothesis.works/articles/rule-based-stateful-testing/ goes into more depth and demonstrates bundles.  Maybe we should link them from the docs?

**Comment by glowthrower** [comment]
I think a link to both at the top of the readthedocs page I originally linked to would be very helpful! 

I remember reading the article about the Die Hard problem years ago when I first started using Hypothesis, but I wouldn't have thought to have gone back to it. 

I wasn't aware that the second article existed -- and it would have been really helpful to me when trying to grok the stateful API, especially since the usage/motivation of bundles in particular is quite a bit clearer and more compelling than that in the readthedocs content. The toy stack code I posted above is a tidied-up version of something I hacked up to experiment with the API concepts, and it's probably unsurprising that in trying to simplify things as much as possible, I wound up with something broadly similar to the first part of the article.

(To put it into a "user story" kind of context, I only started experimenting with this as a result of an idle click-through from [the list of built-in strategies](https://hypothesis.readthedocs.io/en/latest/data.html), which I often have open while I'm writing tests. I don't regularly read through the "main" hypothesis site, since I have a pretty good handle on how the general property-based testing stuff works by now. This might make me an unusual target when considering how to structure the docs...)
